icu4c/source/test/intltest/rbbitst.cpp

```diff
@@ -4883,6 +4897,104 @@ void RBBITest::TestDebugRules() {
+
+    assertEquals("Number of char classes mismatch", numChar + 4, bi->fData->fHeader->fCatCount);
+    assertEquals("Number of states mismatch", numChar + 3, bi->fData->fForwardTable->fNumStates);
+    assertEquals("Trie width mismatch", expectedTrieWidth, ucptrie_getValueWidth(bi->fData->fTrie));
+    assertEquals("Bits of Forward State table mismatch", expectedStateRowBits, bi->fData->fForwardTable->fFlags & RBBI_8BITS_ROWS);
+    assertEquals("Bits of Reverse State table mismatch", expectedStateRowBits, bi->fData->fReverseTable->fFlags & RBBI_8BITS_ROWS);
+
+    bi->setText(text);
+
+    int32_t pos;
+    int32_t i = 0;
+    while ((pos = bi->next()) > 0) {
+        // The firsr numChar should not break between the pair
+        if (i++ < numChar) {
+            assertEquals("next() mismatch", i * 2, pos);
+        } else {
+            // After the first numChar next(), break on each character.
+            assertEquals("next() mismatch", i + numChar, pos);
+        }
+    }
+    while ((pos = bi->previous()) > 0) {
+        // The firsr numChar should not break between the pair
+        if (--i < numChar) {
+            assertEquals("previous() mismatch", i * 2, pos);
+        } else {
+            // After the first numChar next(), break on each character.
+            assertEquals("previous() mismatch", i + numChar, pos);
+        }
+    }
+}
+
+void RBBITest::TestDifferentTrieStateTableBits() {
+    testTrieStateTable(122, UCPTRIE_VALUE_BITS_8, RBBI_8BITS_ROWS);
+    testTrieStateTable(123, UCPTRIE_VALUE_BITS_16, RBBI_8BITS_ROWS);
+    testTrieStateTable(124, UCPTRIE_VALUE_BITS_16, 0 /* 16 bits rows */);
+
+    // Test UCPTRIE_VALUE_BITS_8 with 16 bits rows. Use a different approach to
+    // create state table in 16 bits.
+
+    // Generate 260 'a' as text
+    UnicodeString text("");
+    for (int32_t i = 0; i < 260; i++) {
+        text.append('a');
+    }
+
+    UnicodeString rules("!!quoted_literals_only;'");
+    // 125 'a' in the rule will cause 127 states
+    for (int32_t i = 0; i < 125; i++) {
+        rules.append('a');
+    }
+    rules.append("';.;");
+
+    UErrorCode status = U_ZERO_ERROR;
+    UParseError parseError;
+    LocalPointer<RuleBasedBreakIterator> bi(new RuleBasedBreakIterator(rules, parseError, status));
+
+    assertEquals("Number of states mismatch", 127, bi->fData->fForwardTable->fNumStates);
+    assertEquals("Trie width mismatch", UCPTRIE_VALUE_BITS_8, ucptrie_getValueWidth(bi->fData->fTrie));
+    assertEquals("Bits of Forward State table mismatch", 0, bi->fData->fForwardTable->fFlags & RBBI_8BITS_ROWS);
+    bi->setText(text);
+
+    // break positions:
+    // 125, 250, 251, 252 ... 259
+    assertEquals("next()", 125, bi->next());
+    int32_t i = 0;
+    int32_t pos;
+    while ((pos = bi->next()) > 0) {
+        assertEquals("next()", 250 + i , pos);
+        i++;
+    }
+    i = 0;
+    while ((pos = bi->previous()) > 0) {
```

Comment:
The previous() values will mostly be coming from the break iterator's cache, so this won't be hitting the reverse rules for them. The cache holds 128 recent boundaries.

Reply:
so... how should we test the code for the reverse rules?

Reply:
ping @aheninger

Reply:
Doing a fresh setText() will flush the break cache. But it will start to build up again as soon as you find any boundaries.

Here's a patch to the test I added that will invoke the reverse rules. (verified in the debugger)

```
$ git diff
diff --git a/icu4c/source/test/intltest/rbbitst.cpp b/icu4c/source/test/intltest/rbbitst.cpp
index cced94f1c5..0683d21d6d 100644
--- a/icu4c/source/test/intltest/rbbitst.cpp
+++ b/icu4c/source/test/intltest/rbbitst.cpp
@@ -5044,6 +5044,12 @@ void RBBITest::TestTable_8_16_Bits() {
         bi.setText(testStr);
         assertEquals(WHERE, ruleLen, bi.next());
 
+        // Reverse iteration. Do a setText() first, to flush the break iterator's internal cache
+        // of previously detected boundaries, thus forcing the engine to run the safe reverse rules.
+        bi.setText(testStr);
+        int32_t result = bi.preceding(ruleLen);
+        assertEquals(WHERE, 0, result);
+
         // Verify that the range of rule lengths being tested cover the transations
         // from 8 to 16 bit data.
         bool has8BitRowData = bi.fData->fForwardTable->fFlags & RBBI_8BITS_ROWS;
```

Reply:
done